File: site/backend/src/repositories/rateLimitRepository.ts
import crypto from "node:crypto";
import { storagePaths } from "../config/storagePaths.js";
import { readJsonFile, writeJsonFile } from "../utils/jsonStore.js";

interface RateLimitEntry {
  count: number;
  resetAt: number;
}

type RateLimitStore = Record<string, RateLimitEntry>;

function hashKey(key: string): string {
  return crypto.createHash("sha256").update(key).digest("hex").slice(0, 16);
}

function readStore(): RateLimitStore {
  return readJsonFile<RateLimitStore>(storagePaths.rateLimits, {});
}

function writeStore(store: RateLimitStore): void {
  writeJsonFile(storagePaths.rateLimits, store);
}

function pruneExpired(store: RateLimitStore): RateLimitStore {
  const now = Date.now();
  return Object.fromEntries(
    Object.entries(store).filter(([, entry]) => entry.resetAt > now)
  );
}

export const rateLimitRepository = {
  get(namespace: string, key: string, windowMs: number, maxAttempts: number) {
    const now = Date.now();
    const bucketKey = `${namespace}:${hashKey(key)}`;
    const store = readStore();
    const existing = store[bucketKey];

    if (!existing || existing.resetAt <= now) {
      const entry = { count: 0, resetAt: now + windowMs };
      store[bucketKey] = entry;
      writeStore(pruneExpired(store));
      return { ...entry, remaining: maxAttempts };
    }

    return {
      count: existing.count,
      resetAt: existing.resetAt,
      remaining: Math.max(0, maxAttempts - existing.count),
    };
  },
  hit(namespace: string, key: string, windowMs: number) {
    const now = Date.now();
    const bucketKey = `${namespace}:${hashKey(key)}`;
    const store = readStore();
    const existing = store[bucketKey];

    if (!existing || existing.resetAt <= now) {
      const entry = { count: 1, resetAt: now + windowMs };
      store[bucketKey] = entry;
      writeStore(pruneExpired(store));
      return entry;
    }

    existing.count += 1;
    store[bucketKey] = existing;
    writeStore(pruneExpired(store));
    return existing;
  },
};

File: cms/backend/src/main/java/br/com/rodogarcia/cms/backend/repository/RateLimitRepository.java
package br.com.rodogarcia.cms.backend.repository;

import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Clock;
import java.util.HexFormat;
import java.util.Iterator;

import br.com.rodogarcia.cms.backend.config.StoragePaths;
import org.springframework.stereotype.Repository;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.node.ObjectNode;

@Repository
public class RateLimitRepository {

    private final JsonFileStore store;
    private final Path path;
    private final Clock clock;

    public RateLimitRepository(JsonFileStore store, StoragePaths paths, Clock clock) {
        this.store = store;
        this.path = paths.rateLimits();
        this.clock = clock;
    }

    public RateLimitState get(String namespace, String key, long windowMs, int maxAttempts) {
        return store.withWriteLock(java.util.List.of(path), () -> {
            long now = clock.millis();
            String bucket = namespace + ":" + hashKey(key);
            ObjectNode root = store.readObject(path);
            JsonNode existing = root.get(bucket);
            if (!valid(existing) || existing.path("resetAt").asLong() <= now) {
                ObjectNode entry = store.mapper().createObjectNode();
                entry.put("count", 0);
                entry.put("resetAt", now + windowMs);
                root.set(bucket, entry);
                prune(root, now);
                store.write(path, root);
                return new RateLimitState(0, now + windowMs, maxAttempts);
            }
            int count = existing.path("count").asInt();
            return new RateLimitState(
                count,
                existing.path("resetAt").asLong(),
                Math.max(0, maxAttempts - count)
            );
        });
    }

    /** Verifica o teto e registra o hit na mesma seção crítica. */
    public boolean tryHit(String namespace, String key, long windowMs, int maxAttempts) {
        return store.withWriteLock(java.util.List.of(path), () -> {
            long now = clock.millis();
            String bucket = namespace + ":" + hashKey(key);
            ObjectNode root = store.readObject(path);
            JsonNode existing = root.get(bucket);
            int count = 0;
            long resetAt = now + windowMs;
            if (valid(existing) && existing.path("resetAt").asLong() > now) {
                count = existing.path("count").asInt();
                resetAt = existing.path("resetAt").asLong();
            }
            if (count >= maxAttempts) return false;
            ObjectNode entry = store.mapper().createObjectNode();
            entry.put("count", count + 1);
            entry.put("resetAt", resetAt);
            root.set(bucket, entry);
            prune(root, now);
            store.write(path, root);
            return true;
        });
    }

    private static boolean valid(JsonNode value) {
        return value != null && value.isObject()
            && value.path("count").isNumber() && value.path("resetAt").isNumber();
    }

    private static void prune(ObjectNode root, long now) {
        Iterator<java.util.Map.Entry<String, JsonNode>> entries = root.properties().iterator();
        while (entries.hasNext()) {
            JsonNode value = entries.next().getValue();
            if (!valid(value) || value.path("resetAt").asLong() <= now) entries.remove();
        }
    }

    private static String hashKey(String key) {
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256")
                .digest(key.getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(digest).substring(0, 16);
        } catch (NoSuchAlgorithmException error) {
            throw new IllegalStateException("SHA-256 indisponível.", error);
        }
    }

    public record RateLimitState(int count, long resetAt, int remaining) {
    }
}
